crates/ra_hir/src/diagnostics.rs

```diff
@@ -2,3 +2,53 @@
 pub use hir_def::diagnostics::UnresolvedModule;
 pub use hir_expand::diagnostics::{AstDiagnostic, Diagnostic, DiagnosticSink};
 pub use hir_ty::diagnostics::{MissingFields, MissingMatchArms, MissingOkInTailExpr, NoSuchField};
+
+use std::sync::Arc;
+
+use crate::code_model::Function;
+use crate::db::HirDatabase;
+use crate::has_source::HasSource;
+use hir_ty::{
+    diagnostics::{MissingUnsafe, UnnecessaryUnsafe},
+    expr::unsafe_expressions,
+    InferenceResult,
+};
+use ra_syntax::AstPtr;
+
+pub struct UnsafeValidator<'a, 'b: 'a> {
```

Comment:
I added this validator here rather than in `ra_hir_ty` because I could not access `Function` in order to add the function as the source for the diagnostic messages. Is there a better location for this logic?